ruma-api-macros/src/util.rs

```diff
@@ -261,3 +249,36 @@ pub(crate) fn req_res_name_value<T>(
 pub(crate) fn is_valid_endpoint_path(string: &str) -> bool {
     string.as_bytes().iter().all(|b| (0x21..=0x7E).contains(b))
 }
+
+/// Used by `try_helper` to determine the error and variable to use.
+pub(crate) enum HttpDirection {
+    /// The `RequestDeserializationError` is used and the `request` variable is passed to it's
+    /// constructor.
+    Request,
+    /// The `ResponseDeserializationError` is used and the `response` variable is passed to it's
+    /// constructor.
+    Response,
+}
+
+/// Generates a `match` statement of `expr` that returns early with the correct error based on
+/// the `res_or_req`.
+pub(crate) fn try_helper(expr: TokenStream, res_or_req: HttpDirection) -> TokenStream {
```

Comment:
I don't like this being named `try_helper`, that's too generic of a name. First thing I can come up with is `deser_try`, but I'm also not super happy with that. Do you have ideas for an alternative name?

Reply:
If you have two completely different `quote` invocations for `request` and `response` though, may I also suggest splitting this into two functions (that can then probably live in the `request` and `response` modules)?